keserv/env.go

```diff
@@ -0,0 +1,128 @@
+package keserv
+
+import (
+	"encoding"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/minio/kes"
+)
+
+// Env wraps a type T and an optional environment
+// variable name.
+//
+// It can be used to replace env. variable references
+// with values from the environment during unmarshalling.
+// Further, Env preserves any env. variable references
+// during marshaling.
+//
+// During unmarshalling, Env replaces T's value
+// with the value obtained from the referenced
+// environment variable, if any.
+//
+// During marshaling, Env preserves and encodes
+// the env. variable reference, if not empty.
+// Otherwise, it encodes Env generic type value.
+type Env[T string | kes.Identity | time.Duration] struct {
+	Name  string // Name of the env. variable
+	Value T      // Value obtained by unmarshalling or from the environment
+}
+
+// For now we limit Env[T] to:
+// - string
+// - kes.Identity
+// - time.Duration
+//
+// It is possible to add more types in the future (e.g. net.IP, url.URL, ...)
+// and relax the constraints further (e.g. ~string or ~int64). However,
+// this has implications for marshaling and esp. unmarshaling.
+//
+// For marshaling we can get away with some reasonable interface checks.
+// For example fmt.Stringer and encoding.TextMarshaler.
+// However, unmarshalling semantics are complex. For example, we could allow
+// Env for any type based on int64 via `~int64` - including time.Duration.
+// But we want to unmarshal a time.Duration via time.ParseDuration, not via
+// strconv.ParseInt64. For other types we just don't know whether there is
+// a specialised parse function or whether ParseInt64 should be used.
+//
+// The current implementation is very primitive but good enough
+// for the moment.
+
+func (e Env[_]) MarshalText() ([]byte, error) {
```

Comment:
I thought I knew `Go` until I saw this kind of syntax :sweat:, approved!

Reply:
@Alevsk Say Hi to generics :D